limit.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate
+
+import (
+	"time"
+)
+
+// LimitPer identifies how a limit is allocated.
+type LimitPer string
+
+func (p LimitPer) String() string {
+	return string(p)
+}
+
+// IsValid checks if the given LimitPer is valid.
+func (p LimitPer) IsValid() bool {
+	switch p {
+	case LimitPerTotal, LimitPerIPAddress, LimitPerAuthToken:
+		return true
+	}
+	return false
+}
+
+const (
+	// LimitPerIPAddress indicates that the limit applies per IP address.
+	LimitPerIPAddress LimitPer = "ip-address"
+	// LimitPerAuthToken indicates that the limit applies per auth token.
+	LimitPerAuthToken LimitPer = "auth-token"
+	// LimitPerTotal indicates that the limit applies for all IP address and all Auth Tokens.
+	LimitPerTotal LimitPer = "total"
+)
+
+// Limit defines the number of requests that can be made to perform an action
+// against a resource in a time period, allocated per IP address, auth token,
+// or in total.
+type Limit struct {
+	Resource string
```

Comment:
consider adding docs for each exported field.

Reply:
I will look at this again as I come back to the docs in a number of places. Currently though I am not sure what else to add that isn't covered in the doc for Limit already, but I welcome any suggestions.